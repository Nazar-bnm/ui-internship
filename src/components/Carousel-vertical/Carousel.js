import React, { Component } from 'react';
import cx from 'classnames';
import PropTypes from 'prop-types';

import { scrollTo } from '../../helpers';

import './Carousel.scss';


export const CN = 'carousel-vertical';

class Carousel extends Component {
  constructor(props) {
    super(props);

    this.carouselViewport = React.createRef();
    this.handleClick = this.handleClick.bind(this, props.type);
    this.onResize = this.onResize.bind(this);
    this.onScroll = this.onScroll.bind(this);
    this.countTheSlideWidth = this.countTheSlideWidth.bind(this);
    this.state = {
      numOfSlidesToScroll: 3,
      allTheWayStart: false,
      allTheWayEnd: false,
      slidesCount: this.getSlidesCount()
    };
  }

  componentDidMount() {
    this.resizeTheCarousel();
    window.addEventListener('resize', this.onResize);
    this.checkNumOfSlidesToScroll();
    this.checkIfSlidesAllTheWayOver();
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.onResize);
  }

  onScroll() {
    this.checkIfSlidesAllTheWayOver();
  }

  onResize() {
    this.resizeTheCarousel();
    this.checkNumOfSlidesToScroll();
    this.getSlidesCount();
  }

  getSlidesCount = () => {
    const { visibleNumOfSlides: sizeValues } = this.props;
    let count;

    if (window.innerWidth >= 1024) {
      count = sizeValues.desktop;
    } else if (window.innerWidth >= 768) {
      count = sizeValues.tablet;
    } else if (window.innerWidth >= 320) {
      count = sizeValues.mobile;
    }


    return count;
  };

  resizeTheCarousel() {
    this.setState({ slidesCount: this.getSlidesCount() });
    this.countTheSlideWidth();
  }

  countTheSlideWidth() {
    const { parentElement } = this.carouselViewport.current;
    const { type } = this.props;
    let { clientHeight: clientLength } = parentElement;

    type === 'vertical' && ({ clientHeight: clientLength } = parentElement);
    const { slidesCount } = this.state;

    this.setState({ lengthOfSlide: clientLength / slidesCount });
  }

  checkIfSlidesAllTheWayOver() {
    const { type } = this.props;
    let {
      scrollTop: howMuchIsScrolled,
      clientHeight: clientLength,
      scrollHeight: scrollLength
    } = this.carouselViewport.current;

    if (type === 'vertical') {
      ({
        scrollTop: howMuchIsScrolled,
        clientHeight: clientLength,
        scrollHeight: scrollLength
      } = this.carouselViewport.current);
    }

    // if scrollLeft == 0
    // hide left button
    const allTHeWayStartValue = howMuchIsScrolled === 0;

    // if scrollLeft + viewPortOffset.length === whole viewPort length
    // 9 cards - each 120px: 9 * 210all === whole viewPort length
    // hide the rightScrollButton
    const allTheWayEndValue = howMuchIsScrolled + clientLength === scrollLength;
    const { allTheWayStart, allTheWayEnd } = this.state;

    (allTheWayStart !== allTHeWayStartValue
      || allTheWayEnd !== allTheWayEndValue)
      && this.setState({
        allTheWayStart: allTHeWayStartValue,
        allTheWayEnd: allTheWayEndValue
      });
  }

  checkNumOfSlidesToScroll() {
    const numOfSlidesToScroll = window.innerWidth <= 900 ? 1 : 2;
    const { numOfSlidesToScroll: numOfSlides } = this.state;

    numOfSlides !== numOfSlidesToScroll
      && this.setState({ numOfSlidesToScroll });
  }

  handleClick(type, e) {
    console.log(this.state, '-type   ', e.currentTarget, '-e');
    const clickedBtn = e.currentTarget.classList.contains(`${CN}__left-nav`)
      ? 'left'
      : 'right';
    const { numOfSlidesToScroll, lengthOfSlide } = this.state;
    const carouselViewport = this.carouselViewport.current;
    const step = numOfSlidesToScroll * lengthOfSlide;
    const newPos = clickedBtn === 'left'
      ? carouselViewport.scrollTop - step
      : carouselViewport.scrollTop + step;
    const timeToMoveOneSlide = 200;
    const totalTimetoMove = numOfSlidesToScroll * timeToMoveOneSlide;

    scrollTo({
      element: this.carouselViewport,
      to: newPos,
      duration: totalTimetoMove,
      scrollDirection: 'scrollTop'
    });
  }

  renderChildren() {
    const { children, items } = this.props;
    const { lengthOfSlide } = this.state;
    const style = { maxHeight: `${lengthOfSlide}px` };
    const listOfChildren = [...children];

    return listOfChildren.map((item, i) => (
      <div
        key={items[i].id}
        className={`${CN}__child`}
        style={style}
      >
        {item}
      </div>
    ));
  }


  render() {
    const { allTheWayStart, allTheWayEnd } = this.state;
    const navClasses = cx({ [`${CN}__nav`]: true });
    const leftNavClasses = cx(
      {
        [`${CN}__left-nav`]: true,
        [`${CN}__nav-disabled`]: allTheWayStart
      },
      navClasses
    );
    const rightNavClasses = cx(
      {
        [`${CN}__right-nav`]: true,
        [`${CN}__nav-disabled`]: allTheWayEnd
      },
      navClasses
    );
    return (
      <div className={`${CN} content`}>
        <button
          className={leftNavClasses}
          type="button"
          onClick={this.handleClick}
        >
          <i className={`${CN}__arrow-button chevron left icon`} />
        </button>
        <div
          style={{ flexDirection: this.props.type === 'vertical' && 'column' }}
          className={`${CN}__viewport`}
          ref={this.carouselViewport}
          onScroll={this.onScroll}
        >
          {this.renderChildren()}
        </div>
        <button
          type="button"
          className={rightNavClasses}
          onClick={this.handleClick}
        >
          <i className={`${CN}__arrow-button chevron right icon`} />
        </button>
      </div>
    );
  }
}

Carousel.propTypes = {
  children: PropTypes.array.isRequired
};

export default Carousel;
