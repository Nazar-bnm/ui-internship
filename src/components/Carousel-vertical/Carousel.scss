// 

@import '../../scss/global_styles';

.carousel-vertical {
  border: 1px solid $black;
  position: relative;
  display: flex;
  margin: 30px;
  width: fit-content;
  height: 400px;

  &__viewport {
    display: flex;
    z-index: 1;
    white-space: nowrap;
    overflow-y: hidden;
    overflow-x: hidden; 
    width: fit-content;
    -webkit-overflow-scrolling: touch;
    -ms-overflow-style: none;  // IE 10+
    scrollbar-width: none; // Firefox
  }

  &__viewport::-webkit-scrollbar {
    display: none;  // Safari and Chrome
  }

  &__nav {
    z-index: 3;
    position: absolute;
    display: flex;
    width: 20px;
    height: 35px;
    cursor: pointer;
    padding: 0;
    outline: none;
    border: none;
    background: none;
    opacity: .8;
    transition: opacity .2s ease-in-out;
    color: $black;
  }

  &__left-nav {
    right: 50%;
    top: -35px;
    transform: rotate(90deg);
  }

  &__right-nav {
    right: 50%;
    bottom: -35px;
    transform: rotate(90deg);
  }

  &__nav:hover {
    opacity: 1;
  }

  // &__nav-disabled {
  //   z-index: 0;
  //   opacity: 0;
  //   cursor: default;
  //   user-select: none;
  //   pointer-events: none;
  //   transition: .5;
  // }

  &__child {
    display: inline-block;
    width: 100%;
  }

  i.icon {
    font-size: 1.5rem;
  }
}

@media (max-width: 768px) {
  .carousel {

    &__nav {
      display: none;
    }
  }
}
