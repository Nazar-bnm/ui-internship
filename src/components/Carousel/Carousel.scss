@import '../../scss/global_styles';

.carousel {
  position: relative;
  width: fit-content;
  height: fit-content;

  &-container {
    height: 100%;
    overflow-y: auto;
    overflow-x: auto;
    -webkit-overflow-scrolling: touch;
    -ms-overflow-style: none;
    scrollbar-width: none;

    &::-webkit-scrollbar {
      display: none;
    }
  }

  &__viewport {
    white-space: nowrap;
    width: fit-content;
  }

  .vertical {
    display: flex;
    flex-direction: column;
  }

  &__nav {
    position: absolute;
    cursor: pointer;
    padding: 0;
    outline: none;
    border: none;
    background: none;
    opacity: 0.8;
    transition: opacity 0.2s ease-in-out;
    color: $white;
  }

  &__left-nav {
    left: 1%;
    top: 50%;
    transform: translateY(-50%);
  }

  &__right-nav {
    right: 1%;
    top: 50%;
    transform: translateY(-50%);
  }

  &__top-nav {
    top: 1%;
    left: 50%;
    transform: translateX(-50%);
  }

  &__bottom-nav {
    bottom: 1%;
    left: 50%;
    transform: translateX(-50%);
  }

  &__nav:hover {
    opacity: 1;
  }

  &__nav-disabled {
    z-index: 0;
    opacity: 0;
    cursor: default;
    user-select: none;
    pointer-events: none;
    transition: 0.5;
  }

  &__child {
    display: inline-block;
    text-align: center;
    width: 100%;
    height: 100%;
  }

  i.carousel__arrow-button {
    opacity: 0.4;
    margin: 0;
    width: 2rem;
    height: 2rem;
    font-size: 2rem;
    border-radius: 50%;
    background-color: $emperor;
    color: $alto;
    cursor: pointer;

    &:hover,
    &:active {
      opacity: 0.6;
    }
  }
}

@media (max-width: 768px) {
  .carousel {
    &__nav {
      display: none;
    }
  }
}