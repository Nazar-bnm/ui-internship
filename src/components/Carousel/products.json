[
  {
    "productName": "Tommy Jeans Shirt",
    "description": "Shirt from the Tommy Jeans collection. The model is made of smooth fabric. Has a classic collar.\n      - Regular cut.\n      - Soft classic Kent collar.\n      - Button fastening.\n      - Thin, inflexible fabric.\n      - Collar circumference: 40.5 cm.\n      - Sleeve length: 67 cm.\n      - Length: 78 cm.\n      - Armpit width: 54 cm.\n      - Shoulder width: 42.5 cm.\n      - Dimensions for size: M.\n\n      Pattern: smooth\n      Cut: slim\n      Cut: regular\n      Material: cotton\n      Collar: classic\n      Collar: soft\n      Occasions: casual\n      Composition: 100% Cotton",
    "brand": "Tommy Jeans",
    "sizes": [
      "s",
      "m",
      "l",
      "xl"
    ],
    "colors": [
      "#a8c8e0",
      "#ffffff",
      "#012961"
    ],
    "images": [
      {
        "value": "#a8c8e0",
        "url": [
          "src/assets/img/content/products/shirts/tommy-blue-1.1.jpg",
          "tommy-blue-1.2.jpg",
          "tommy-blue-1.3.jpg",
          "tommy-blue-1.4.jpg"
        ]
      },
      {
        "value": "#ffffff",
        "url": [
          "tommy-white-1.1.jpg",
          "tommy-white-1.2.jpg",
          "tommy-white-1.3.jpg",
          "tommy-white-1.4.jpg"
        ]
      },
      {
        "value": "#012961",
        "url": [
          "tommy-darkblue-1.1.jpg",
          "tommy-darkblue-1.2.jpg",
          "tommy-darkblue-1.3.jpg",
          "tommy-darkblue-1.4.jpg"
        ]
      }
    ],
    "video": "./video/tommy-all.mp4",
    "sex": "man",
    "category": "shirt",
    "season": [
      "fall",
      "spring"
    ],
    "quantity": 10,
    "price": 69
  },
  {
    "productName": "Armani Exchange - Jeans Shirt",
    "description": "Shirt from the Armani Exchange collection. The model is made of smooth fabric. Has an Italian collar.\n\n    - Simple cut.\n    - Button fastening.\n    - Sleeves with buttoned cuffs.\n    - Smooth fabric.\n\n    Pattern: smooth\n    Cut: regular\n    Collar: soft\n    Occasions: casual\n    Collar: Italian",
    "brand": "Armani Exchange",
    "sizes": [
      "s",
      "m",
      "l",
      "xl"
    ],
    "colors": [
      "#445d80"
    ],
    "images": [
      {
        "value": "#445d80",
        "url": [
          "src/assets/img/content/products/shirts/armani-jeansblue-1.1.jpg",
          "armani-jeansblue-1.2.jpg",
          "armani-jeansblue-1.3.jpg",
          "armani-jeansblue-1.4.jpg"
        ]
      }
    ],
    "video": "",
    "sex": "man",
    "category": "shirt",
    "season": [
      "fall",
      "spring"
    ],
    "quantity": 10,
    "price": 79
  },
  {
    "productName": "Polo Ralph Lauren (smooth)",
    "description": "Shirt from the Polo Ralph Lauren collection. The model is made of smooth fabric.\n        - Slim cut.\n        - Sleeves with buttoned cuffs.\n        - Soft button-down collar.\n        - Button fastening.\n        - Smooth fabric.\n        - Sleeve length: 66 cm.\n        - Length: 76 cm.\n        - Armpit width: 52 cm.\n        - Shoulder width: 43 cm.\n        - Dimensions for size: M.\n\n        Pattern: smooth\n        Cut: slim\n        Material: cotton\n        Collar: soft\n        Occasions: casual\n        Collar: with a button\n        Composition:100% Cotton.",
    "brand": "Polo Ralph Lauren",
    "sizes": [
      "s",
      "m",
      "l",
      "xl"
    ],
    "colors": [
      "#000000",
      "#ffffff",
      "#00064d"
    ],
    "images": [
      {
        "value": "#ffffff",
        "url": [
          "src/assets/img/content/products/shirts/polo-rl-white-1.1.jpg",
          "polo-rl-white-1.2.jpg",
          "polo-rl-white-1.3.jpg",
          "polo-rl-white-1.4.jpg"
        ],
        "video": "./video/polo-rl-white.mp4"
      },
      {
        "value": "#00064d",
        "url": [
          "polo-rl-blue-1.1.jpg",
          "polo-rl-blue-1.2.jpg",
          "polo-rl-blue-1.3.jpg",
          "polo-rl-blue-1.4.jpg"
        ]
      }
    ],
    "video": "./video/polo-rl-blue.mp4",
    "sex": "man",
    "category": "shirt",
    "season": [
      "fall",
      "spring"
    ],
    "quantity": 8,
    "price": 129
  },
  {
    "productName": "SCOTCH & SODA - Shirt",
    "description": "Shirt from the Scotch & Soda collection. The model is made of patterned fabric. It has a classic, soft collar.\n      - Regular cut.\n      - Soft classic Kent collar.\n      - Sleeves with buttoned cuffs.\n      - Chest pocket.\n      - Patterned fabric.\n      - Armpit width: 50 cm.\n      - Collar circumference: 41 cm.\n      - Shoulder width: 43 cm.\n      - Sleeve length: 67 cm.\n      - Length: 71 cm.\n      - Dimensions for size: M.\n      \n      Cut: regular\n      Pattern: small pattern\n      Material: cotton\n      Collar: classic\n      Collar: soft\n      Occasions: casual\n\n      Composition: 100% Cotton.",
    "brand": "SCOTCH & SODA",
    "sizes": [
      "s",
      "m",
      "l",
      "xl"
    ],
    "colors": [
      "#013005",
      "#000000",
      "#3d3d3d"
    ],
    "images": [
      {
        "value": "#000000",
        "url": [
          "src/assets/img/content/products/shirts/scotch-soda-black-1.1.jpg",
          "scotch-soda-black-1.2.jpg",
          "scotch-soda-black-1.3.jpg",
          "scotch-soda-black-1.4.jpg"
        ]
      },
      {
        "value": "#3d3d3d",
        "url": [
          "scotch-soda-grey-1.1.jpg",
          "scotch-soda-grey-1.2.jpg",
          "scotch-soda-grey-1.3.jpg",
          "scotch-soda-grey-1.4.jpg"
        ]
      }
    ],
    "video": "./video/scotch-soda-green.mp4",
    "sex": "man",
    "category": "shirt",
    "season": [
      "fall",
      "spring"
    ],
    "quantity": 8,
    "price": 109
  },
  {
    "productName": "LEVI`S - Shirt",
    "description": "Shirt from the Levi's collection. The model is made of smooth fabric.\n    - Soft classic Kent collar.\n    - Button fastening.\n    - Soft cuffs.\n    - Thick, elastic fabric.\n    - Smooth material.\n    - Collar circumference: 42 cm.\n    - Length: 70 cm.\n    - Armpit width: 54 cm.\n    - Shoulder width: 46 cm.\n    - Dimensions for size: M.\n\n    Composition: 100% Cotton",
    "brand": "LEVI`S",
    "sizes": [
      "s",
      "m",
      "l",
      "xl"
    ],
    "colors": [
      "#023d0b"
    ],
    "images": [
      {
        "value": "#023d0b",
        "url": [
          "src/assets/img/content/products/shirts/levis-sol-green-1.1.jpg",
          "levis-sol-green-1.2.jpg",
          "levis-sol-green-1.3.jpg",
          "levis-sol-green-1.4.jpg"
        ]
      }
    ],
    "video": "./video/levis-sol-green.mp4",
    "sex": "man",
    "category": "shirt",
    "season": [
      "fall",
      "spring"
    ],
    "quantity": 15,
    "price": 79
  },
  {
    "productName": "LEVI`S - Checked Shirt",
    "description": "Shirt from the Levi's collection. The model is made of smooth fabric.\n      - Slim cut.\n      - Snap closure.\n      - Soft classic Kent collar.\n      - Chest pockets.\n      - Collar circumference: 44 cm.\n      - Sleeve length: 64 cm.\n      - Length: 71 cm.\n      - Armpit width: 53 cm.\n      - Shoulder width: 44 cm.\n      - Dimensions for size: M.\n\n      Pattern: patterned\n      Cut: slim\n      Material: cotton\n      Collar: classic\n      Design: lattice\n      Collar: soft\n      Occasions: casual\n      Composition: 100% Cotton",
    "brand": "LEVI`S",
    "sizes": [
      "s",
      "m",
      "l",
      "xl"
    ],
    "colors": [
      "#245eff",
      "#d41919"
    ],
    "images": [
      {
        "value": "#d41919",
        "url": [
          "src/assets/img/content/products/shirts/levis-red-1.1.jpg",
          "levis-red-1.2.jpg",
          "levis-red-1.3.jpg",
          "levis-red-1.4.jpg"
        ]
      }
    ],
    "video": "./video/levis-red.mp4",
    "sex": "man",
    "category": "shirt",
    "season": [
      "fall",
      "spring"
    ],
    "quantity": 10,
    "price": 79
  },
  {
    "productName": "Polo Ralph Lauren - Shirt",
    "description": "Shirt from the Polo Ralph Lauren collection. The model is made of smooth fabric. It has an Italian, slightly stiffened collar.\n\n      - Slim cut.\n      - Slightly stiffened Italian collar.\n      - Button fastening.\n      - Thin, inflexible fabric.\n      - Collar circumference: 39.5 cm.\n      - Armpit width: 53.5 cm.\n      - Shoulder width: 43 cm.\n      - Sleeve length: 66 cm.\n      - Length: 75.5 cm.\n      - Dimensions for size: M.\n      \n      Pattern: smooth\n      Cut: slim\n      Material: cotton\n      Occasions: casual\n      Collar: Italian\n      Collar: stiffened\n      Composition: 100% Cotton",
    "brand": "Polo Ralph Lauren",
    "sizes": [
      "s",
      "m",
      "l",
      "xl"
    ],
    "colors": [
      "#98c3f5",
      "#002045"
    ],
    "images": [
      {
        "value": "#98c3f5",
        "url": [
          "src/assets/img/content/products/shirts/polorl-lightblue-1.1.jpg",
          "polorl-lightblue-1.1.jpg",
          "polorl-lightblue-1.1.jpg",
          "polorl-lightblue-1.1.jpg"
        ]
      },
      {
        "value": "#002045",
        "url": [
          "polorl-darkblue-1.1.jpg",
          "polorl-darkblue-1.1.jpg",
          "polorl-darkblue-1.1.jpg",
          "polorl-darkblue-1.1.jpg"
        ]
      }
    ],
    "video": "./video/polorl-lightblue.mp4",
    "sex": "man",
    "category": "shirt",
    "season": [
      "fall",
      "spring"
    ],
    "quantity": 20,
    "price": 59
  },
  {
    "productName": "Medicine - Checked Shirt",
    "description": "Shirt from the Medicine collection. The model is made of plaid fabric. It has a classic, soft collar.\n\n    - Regular cut.\n    - Button fastening.\n    - Sleeves with buttoned cuffs.\n    - Soft classic Kent collar.\n    - Patterned fabric.\n    - Collar circumference: 43 cm.\n    - Armpit width: 53 cm.\n    - Shoulder width: 43 cm.\n    - Sleeve length: 66 cm\n    - Length: 75 cm.\n    - Dimensions for size: M.\n    \n    Pattern: patterned\n    Cut: regular\n    Material: cotton\n    Collar: classic\n    Design: lattice\n    Collar: soft\n    Occasions: casual\n    Composition: 100% Cotton",
    "brand": "Medicine",
    "sizes": [
      "s",
      "m",
      "l",
      "xl"
    ],
    "colors": [
      "#ffbf00",
      "#c70202",
      "#696060"
    ],
    "images": [
      {
        "value": "#c70202",
        "url": [
          "src/assets/img/content/products/shirts/medicine-red-1.1.jpg",
          "medicine-red-1.2.jpg",
          "medicine-red-1.3.jpg",
          "medicine-red-1.4.jpg"
        ]
      },
      {
        "value": "#696060",
        "url": [
          "medicine-gray-1.1.jpg",
          "medicine-gray-1.2.jpg",
          "medicine-gray-1.3.jpg",
          "medicine-gray-1.4.jpg"
        ]
      }
    ],
    "video": "./video/medicine-all.mp4",
    "sex": "man",
    "category": "shirt",
    "season": [
      "fall",
      "spring"
    ],
    "quantity": 20,
    "price": 59
  },
  {
    "productName": "Armany Exchange - Shirt",
    "description": "Shirt from the Armani Exchange collection. The model is made of smooth fabric. Has an Italian collar.\n\n      - Simple cut.\n      - Button fastening.\n      - Sleeves with buttoned cuffs.\n      - Smooth fabric.\n\n      Pattern: smooth\n      Cut: regular\n      Collar: soft\n      Occasions: casual\n      Collar: Italian",
    "brand": "Armani Exchange",
    "sizes": [
      "s",
      "m",
      "l",
      "xl"
    ],
    "colors": [
      "#ffffff",
      "#000000"
    ],
    "images": [
      {
        "value": "#000000",
        "url": [
          "src/assets/img/content/products/shirts/armani-black-1.1.jpg",
          "armani-black-1.2.jpg",
          "armani-black-1.3.jpg",
          "armani-black-1.4.jpg"
        ]
      }
    ],
    "video": "./video/armani-all.mp4",
    "sex": "man",
    "category": "shirt",
    "season": [
      "fall",
      "spring"
    ],
    "quantity": 20,
    "price": 59
  }
]