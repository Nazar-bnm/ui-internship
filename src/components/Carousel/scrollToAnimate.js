function scrollTo(params) {
  const { element, to, duration, scrollDirection } = params;
  const start = element.current[scrollDirection];
  const change = to - start;
  const increment = ( 1000 / 50 );

  const animatedScroll = (elapsedTime) => {
    elapsedTime += increment;
    const position = easeInOut(elapsedTime, start, change, duration);
    element.current[scrollDirection] = position;
    if (elapsedTime < duration) {
      window.requestAnimationFrame(animatedScroll.bind(null, elapsedTime));
      // setTimeout(function() {
      //   animatedScroll(elapsedTime)
      // }, increment);
    }
  };

  // animatedScroll(0);
  window.requestAnimationFrame(animatedScroll.bind(null, 0));
}

function easeInOut(currentTime, start, change, duration) {
  currentTime /= duration / 2;

  if (currentTime < 1) {
    return change / 2 * currentTime * currentTime + start;
  }
  currentTime -= 1;
  return -change / 2 * (currentTime * (currentTime - 2) - 1) + start;
}

export default scrollTo;
