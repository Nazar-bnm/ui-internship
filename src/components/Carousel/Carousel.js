import React, { Component } from 'react';
import cx from 'classnames';
import PropTypes from 'prop-types';

import { scrollTo } from '../../helpers';

import './Carousel.scss';


export const CN = 'carousel';

class Carousel extends Component {
  constructor(props) {
    super(props);

    this.carouselViewport = React.createRef();
    this.handleClick = this.handleClick.bind(this);
    this.onResize = this.onResize.bind(this);
    this.onScroll = this.onScroll.bind(this);
    this.countTheSlideWidth = this.countTheSlideWidth.bind(this);
    this.state = {
      numOfSlidesToScroll: 3,
      allTheWayLeft: false,
      allTheWayRight: false,
      slidesCount: this.getSlidesCount()
    };
  }

  componentDidMount() {
    this.resizeTheCarousel();
    window.addEventListener('resize', this.onResize);
    this.checkNumOfSlidesToScroll();
    this.checkIfSlidesAllTheWayOver();
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.onResize);
  }

  onScroll() {
    this.checkIfSlidesAllTheWayOver();
  }

  onResize() {
    this.resizeTheCarousel();
    this.checkNumOfSlidesToScroll();
    this.getSlidesCount();
  }

  getSlidesCount = () => {
    const sizeValues = {
      desktop: 3,
      tablet: 2,
      mobile: 1
    };

    let count;

    if (window.innerWidth >= 1024) {
      count = sizeValues.desktop;
    } else if (window.innerWidth >= 768) {
      count = sizeValues.tablet;
    } else if (window.innerWidth >= 320) {
      count = sizeValues.mobile;
    }

    return count;
  };

  resizeTheCarousel() {
    this.setState({ slidesCount: this.getSlidesCount() });
    this.countTheSlideWidth();
  }

  countTheSlideWidth() {
    const { scrollWidth } = this.carouselViewport.current.parentElement;
    const { slidesCount } = this.state;

    this.setState({ widthOfSlide: scrollWidth / slidesCount });
  }

  checkIfSlidesAllTheWayOver() {
    const {
      scrollLeft,
      clientWidth,
      scrollWidth
    } = this.carouselViewport.current;
    // if scrollLeft == 0
    // hide left button
    const allTHeWayLeftValue = scrollLeft === 0;

    // if scrollLeft + viewPortOffset.length === whole viewPort length
    // 9 cards - each 120px: 9 * 210all === whole viewPort length
    // hide the rightScrollButton
    const allTheWayRightValue = scrollLeft + clientWidth === scrollWidth;
    const { allTheWayLeft, allTheWayRight } = this.state;

    (allTheWayLeft !== allTHeWayLeftValue
      || allTheWayRight !== allTheWayRightValue)
      && this.setState({
        allTheWayLeft: allTHeWayLeftValue,
        allTheWayRight: allTheWayRightValue
      });
  }

  checkNumOfSlidesToScroll() {
    const numOfSlidesToScroll = window.innerWidth <= 900 ? 1 : 2;
    const { numOfSlidesToScroll: numOfSlides } = this.state;

    numOfSlides !== numOfSlidesToScroll
      && this.setState({ numOfSlidesToScroll });
  }

  handleClick(e) {
    const clickedBtn = e.currentTarget.classList.contains(`${CN}__left-nav`)
      ? 'left'
      : 'right';
    const { numOfSlidesToScroll, widthOfSlide } = this.state;
    const carouselViewport = this.carouselViewport.current;
    const step = numOfSlidesToScroll * widthOfSlide;
    const newPos = clickedBtn === 'left'
      ? carouselViewport.scrollLeft - step
      : carouselViewport.scrollLeft + step;
    const timeToMoveOneSlide = 200;
    const totalTimetoMove = numOfSlidesToScroll * timeToMoveOneSlide;

    scrollTo({
      element: this.carouselViewport,
      to: newPos,
      duration: totalTimetoMove,
      scrollDirection: 'scrollLeft'
    });
  }

  renderChildren() {
    const { children, items } = this.props;
    const { widthOfSlide } = this.state;
    const style = { maxWidth: `${widthOfSlide}px` };
    const listOfChildren = [...children];

    return listOfChildren.map((item, i) => (
      <div
        key={items[i].id}
        className={`${CN}__child`}
        style={style}
      >
        {item}
      </div>
    ));
  }


  render() {
    const { allTheWayLeft, allTheWayRight } = this.state;
    const navClasses = cx({ [`${CN}__nav`]: true });
    const leftNavClasses = cx(
      {
        [`${CN}__left-nav`]: true,
        [`${CN}__nav-disabled`]: allTheWayLeft
      },
      navClasses
    );
    const rightNavClasses = cx(
      {
        [`${CN}__right-nav`]: true,
        [`${CN}__nav-disabled`]: allTheWayRight
      },
      navClasses
    );

    return (
      <div className={`${CN} content`}>
        <button
          className={leftNavClasses}
          type="button"
          onClick={this.handleClick}
        >
          <i className={`${CN}__arrow-button chevron left icon`} />
        </button>
        <div
          className={`${CN}__viewport`}
          ref={this.carouselViewport}
          onScroll={this.onScroll}
        >
          {this.renderChildren()}
        </div>
        <button
          type="button"
          className={rightNavClasses}
          onClick={this.handleClick}
        >
          <i className={`${CN}__arrow-button chevron right icon`} />
        </button>
      </div>
    );
  }
}

Carousel.propTypes = {
  children: PropTypes.array.isRequired
};

export default Carousel;
