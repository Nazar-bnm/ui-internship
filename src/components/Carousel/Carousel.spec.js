import React from 'react';
import { mount } from 'enzyme';

import Carousel from './Carousel';

describe('<Carousel />', () => {
  let items;
  let wrapper;
  
  beforeEach( () => {
    items = [
      <div id={1}><div key={1}>hello1 </div></div>,
      <div id={2}><div key={2}>hello 2</div></div>,
      <div id={3}><div key={3}>hello3 </div></div>,
      <div id={4}><div key={4}>hello 4</div></div>,
      <div id={5}><div key={5}>hello 5</div></div>,
      <div id={6}><div key={6}>hello 6</div></div>,
      <div id={7}><div key={7}>hello 7</div></div>,
    ];
    wrapper = mount(<Carousel>{items}</Carousel>);
  });

  test('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should check `componentWillUnmount()`', () => {
    const inst = wrapper.instance();
    const spy = jest.spyOn(inst, 'componentWillUnmount');

    wrapper.unmount();
    expect(spy).toHaveBeenCalled();
  });

  test('should check `onScroll()`', () => {
    const inst = wrapper.instance();
    const spy = jest.spyOn(inst, 'onScroll');

    inst.onScroll();
    expect(spy).toHaveBeenCalled();
  });

  test('should check `onResize()`', () => {
    const inst = wrapper.instance();
    const spy = jest.spyOn(inst, 'onResize');

    inst.onResize();
    expect(spy).toHaveBeenCalled();
  });

  test('should check `handleClick()`', () => {
    jest.spyOn(Carousel.prototype, 'handleClick');
    wrapper = mount(<Carousel>{items}</Carousel>);

    const carouselNavs = wrapper.find('.carousel__nav');
    const leftNav = carouselNavs.at(0);
    const rightNav = carouselNavs.at(1);

    leftNav.simulate('click');
    rightNav.simulate('click');
    expect(Carousel.prototype.handleClick).toHaveBeenCalledTimes(2);
  });
});