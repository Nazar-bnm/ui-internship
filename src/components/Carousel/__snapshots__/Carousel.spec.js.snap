// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Carousel /> should match snapshot 1`] = `
<Carousel>
  <div
    className="carousel content content__mobile-width"
  >
    <button
      className="carousel__nav carousel__left-nav carousel__nav-disabled"
      onClick={[Function]}
      type="button"
    >
      <i
        className="carousel__arrow-button chevron left icon"
      />
    </button>
    <div
      className="carousel__viewport"
      onScroll={[Function]}
    >
      <div
        className="carousel__child"
        key="1"
        style={
          Object {
            "maxWidth": undefined,
          }
        }
      >
        <div
          key="1"
        >
          <div
            key="1"
          >
            hello1 
          </div>
        </div>
      </div>
      <div
        className="carousel__child"
        key="2"
        style={
          Object {
            "maxWidth": undefined,
          }
        }
      >
        <div
          key="2"
        >
          <div
            key="2"
          >
            hello 2
          </div>
        </div>
      </div>
      <div
        className="carousel__child"
        key="3"
        style={
          Object {
            "maxWidth": undefined,
          }
        }
      >
        <div
          key="3"
        >
          <div
            key="3"
          >
            hello3 
          </div>
        </div>
      </div>
      <div
        className="carousel__child"
        key="4"
        style={
          Object {
            "maxWidth": undefined,
          }
        }
      >
        <div
          key="4"
        >
          <div
            key="4"
          >
            hello 4
          </div>
        </div>
      </div>
      <div
        className="carousel__child"
        key="5"
        style={
          Object {
            "maxWidth": undefined,
          }
        }
      >
        <div
          key="5"
        >
          <div
            key="5"
          >
            hello 5
          </div>
        </div>
      </div>
      <div
        className="carousel__child"
        key="6"
        style={
          Object {
            "maxWidth": undefined,
          }
        }
      >
        <div
          key="6"
        >
          <div
            key="6"
          >
            hello 6
          </div>
        </div>
      </div>
      <div
        className="carousel__child"
        key="7"
        style={
          Object {
            "maxWidth": undefined,
          }
        }
      >
        <div
          key="7"
        >
          <div
            key="7"
          >
            hello 7
          </div>
        </div>
      </div>
    </div>
    <button
      className="carousel__right-nav carousel__nav carousel__nav-disabled"
      onClick={[Function]}
      type="button"
    >
      <i
        className="carousel__arrow-button chevron right icon"
      />
    </button>
  </div>
</Carousel>
`;
